## Project 4 - Document Classifier Most Informative Features

#### Using the movie review classifier discussed in this chapter, generate a list of the thirty features the classifier finds to be most informative. Can you explain why these features are informative? Do you find any of them surprising?

In order to solve this I adapted the movie document classifier code in a class called MovieClassifier. On instantiation it builds automatically parses and builds featuresets from the movie review corpus. A function is then passed in to the public train method to extract features. A call to the most informative features wrapper returns n number of important features for the training data 


In [2]:
from movie_classifier import MovieClassifier

In [4]:
m = MovieClassifier()

Below is the feature function defined in the text to classify the movie reviews

In [5]:
def document_features(document, word_features):
        document_words = set(document)
        features = {}
        for w in word_features:
            features['contains({})'.format(w)] = (w in document_words)
        return features

In [6]:
m.train(document_features)

After the model is trained with NaiveBayesClassifier method we can look at the top 30 most important features

In [7]:
m.most_informative(30)

Most Informative Features
          contains(sans) = True              neg : pos    =      9.2 : 1.0
    contains(mediocrity) = True              neg : pos    =      7.8 : 1.0
     contains(dismissed) = True              pos : neg    =      6.9 : 1.0
        contains(fabric) = True              pos : neg    =      6.2 : 1.0
   contains(overwhelmed) = True              pos : neg    =      6.2 : 1.0
         contains(wires) = True              neg : pos    =      5.8 : 1.0
     contains(uplifting) = True              pos : neg    =      5.7 : 1.0
       contains(topping) = True              pos : neg    =      5.5 : 1.0
          contains(wits) = True              pos : neg    =      5.5 : 1.0
           contains(ugh) = True              neg : pos    =      5.5 : 1.0
  contains(effortlessly) = True              pos : neg    =      5.5 : 1.0
   contains(bruckheimer) = True              neg : pos    =      5.1 : 1.0
          contains(lang) = True              pos : neg    =      4.9 : 1.0

The results show an interesting mix of adjectives, subjects and names. The top most informative feature is the word 'sans', which has a common meaning of 'lacking' or 'without'. A review is 9.2 times more likely to be negative if this word is present in the document. A review with the word 'mediocrity' is 7.8 times more likely to be negative.

There are also several interesting names in the features list. The words 'bruckheimer', 'cronenberg' and 'matheson' refer to directors George Bruckheimer, David Cronenberg and Richard Matheson. In general these directors produce well received movies and reviews are 3 to 4 times more likely to be positive if their names are present. The word 'hugo' is also positive and most likely means that the movie recieved a 'Hugo' award. 

Finally we can test the classifier for accuracy.

In [8]:
m.report_accuracy()

0.71

This means that the document classifier is correct in its prediction 71% of the time